internal/schema/schema.go

```diff
@@ -183,24 +183,39 @@ func (i GetIndexDefStatement) ToCreateIndexConcurrently() (string, error) {
 	return idxToConcurrentlyRegex.ReplaceAllString(string(i), "${1}CONCURRENTLY ${3}"), nil
 }
 
-type Index struct {
-	TableName string
-	Name      string
-	Columns   []string
-	IsInvalid bool
-	IsPk      bool
-	IsUnique  bool
-	// ConstraintName is the name of the constraint associated with an index. Empty string if no associated constraint.
-	// Once we need support for constraints not associated with indexes, we'll add a
-	// Constraint schema object and starting fetching constraints directly
-	ConstraintName string
+type (
+	IndexConstraintType string
+
+	// IndexConstraint informally represents a constraint that is always 1:1 with an index, i.e.,
+	// primary and unique constraints. It's easiest to just treat these like a property of the index rather than
+	// a separate entity
+	IndexConstraint struct {
+		Type                  IndexConstraintType
+		EscapedConstraintName string
+		ConstraintDef         string
+		IsLocal               bool
+	}
 
-	// GetIndexDefStmt is the output of pg_getindexdef
-	GetIndexDefStmt GetIndexDefStatement
+	Index struct {
+		TableName string
+		Name      string
+		Columns   []string
+		IsInvalid bool
+		IsUnique  bool
 
-	// ParentIdxName is the name of the parent index if the index is a partition of an index
-	ParentIdxName string
-}
+		Constraint *IndexConstraint
+
+		// GetIndexDefStmt is the output of pg_getindexdef
+		GetIndexDefStmt GetIndexDefStatement
+
+		// ParentIdxName is the name of the parent index if the index is a partition of an index
+		ParentIdxName string
+	}
+)
+
+const (
+	PkIndexConstraintType IndexConstraintType = "PRIMARY KEY"
```

Comment:
thought: Should IndexConstraintType and PkIndexConstraintType be public if IsPk is public? Seems like the current pattern so no need to change for this PR.

Reply:
We currently use it to generate the constraint definition. This should be useful when we add support for unique constraints